pkg/package-server-manager/controller.go

```diff
@@ -0,0 +1,148 @@
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	infrastructureName  = "cluster"
+	defaultReplicaCount = int32(2)
+	defaultRolloutCount = 1
+	singleReplicaCount  = int32(1)
+)
+
+// PackageServerCSVReconciler reconciles the PackageServer CSV object
+type PackageServerCSVReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	Lock   sync.Mutex
+
+	Name      string
+	Namespace string
+	Image     string
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which is responsible
+// for ensuring that the PackageServer CSV is deployed in a HA or non-HA mode depending
+// on the topology mode exposed by the current state of the Infrastructure cluster
+// singleton resource.
+func (r *PackageServerCSVReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("csv", req.NamespacedName)
+
+	log.Info("handling current request", "request", req.String())
+	defer log.Info("finished request reconciliation")
+
+	var infra configv1.Infrastructure
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: infrastructureName}, &infra); err != nil {
+		return ctrl.Result{}, err
+	}
+	highAvailabilityMode := getTopologyModeFromInfra(&infra)
+	log.Info("currently topology mode", "highly available", highAvailabilityMode)
+
+	required, err := manifests.NewPackageServerCSV(
+		manifests.WithName(r.Name),
+		manifests.WithNamespace(r.Namespace),
+		manifests.WithImage(r.Image),
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	res, err := controllerutil.CreateOrUpdate(ctx, r.Client, required, func() error {
+		return reconcileCSV(r.Log, r.Image, required, highAvailabilityMode)
+	})
+	if err != nil {
+		log.Error(err, "failed to create or update the csv")
+		return ctrl.Result{}, nil
+	}
+	log.Info("reconciliation result", "res", res)
+
+	return ctrl.Result{}, nil
+}
+
+func reconcileCSV(log logr.Logger, image string, csv *olmv1alpha1.ClusterServiceVersion, highAvailabilityMode bool) error {
+	if csv.ObjectMeta.CreationTimestamp.IsZero() {
+		log.Info("attempting to create the packageserver csv")
+	}
+	if !validateCSV(log, csv) {
+		log.Info("updating invalid csv to use the default configuration")
+		tmp, err := manifests.NewPackageServerCSV(
+			manifests.WithName(csv.Name),
+			manifests.WithNamespace(csv.Namespace),
+			manifests.WithImage(image),
+		)
+		if err != nil {
+			return err
+		}
+		csv.Spec = tmp.Spec
+	}
+	if !ensureCSV(log, image, csv, highAvailabilityMode) {
+		log.V(3).Info("no further updates are necessary to the csv")
```

Comment:
nit: might want to indicate this log is for packageserver (or omit it altogether)